<html>     <head>         <title>rnaQUAST 2.2.1 Manual</title>         <style type="text/css">             .highlight  pre {               background-color: #f0f2f4;               border-radius: 2px;               font-size: 100%;               line-height: 1.45;               overflow: auto;               padding: 16px;             }         </style>     </head>     <body><h1>rnaQUAST 2.2.1 manual</h1>
<p>1. <a href="#sec1">About rnaQUAST</a><br>
2. <a href="#sec2">Installation &amp; requirements</a><br>
    2.1. <a href="#sec2.1">General requirements</a><br>
    2.2. <a href="#sec2.2">Software for <em>de novo</em> quality assessments</a><br>
    2.3. <a href="#sec2.3">Read alignment software</a><br>
3. <a href="#sec3">Options</a><br>
    3.1. <a href="#sec3.1">Input data options</a><br>
    3.2. <a href="#sec3.2">Basic options</a><br>
    3.3. <a href="#sec3.3">Advanced options</a><br>
4. <a href="#sec4">Understanding rnaQUAST output</a><br>
    4.1. <a href="#sec4.1">Reports</a><br>
    4.2. <a href="#sec4.2">Detailed output</a><br>
    4.3. <a href="#sec4.3">Plots</a><br>
5. <a href="#sec5">Citation</a><br>
6. <a href="#sec6">Feedback and bug reports</a></p>
<p><a name="sec1"></a></p>
<h2>1 About rnaQUAST</h2>
<p>rnaQUAST is a tool for evaluating RNA-Seq assemblies using reference genome and gene database. In addition, rnaQUAST is also capable of estimating gene database coverage by raw reads and <em>de novo</em> quality assessment using third-party software.</p>
<p>rnaQUAST version 2.2.1 was released under GPLv2 on October 14th, 2021 and can be downloaded from <a href="http://cab.spbu.ru/software/rnaquast/" rel="nofollow">http://cab.spbu.ru/software/rnaquast/</a> or <a href="https://github.com/ablab/rnaquast/releases">https://github.com/ablab/rnaquast/releases</a>.</p>
<p><strong>For impatient people:</strong></p>
<ul>
<li>
<p>You will need Python, <a href="https://pythonhosted.org/gffutils/installation.html" rel="nofollow">gffutils</a>, <a href="http://matplotlib.org/" rel="nofollow">matplotlib</a> and <a href="https://joblib.readthedocs.io/en/latest/" rel="nofollow">joblib</a>. Also you will need <a href="http://research-pub.gene.com/gmap/" rel="nofollow">GMAP</a> (or <a href="http://hgwdev.cse.ucsc.edu/~kent/exe/" rel="nofollow">BLAT</a>) and <a href="http://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/" rel="nofollow">BLASTN</a> installed on your machine and added to the <code>$PATH</code> variable.</p>
</li>
<li>
<p>You may also install rnaQUAST via conda</p>
<pre><code> conda install -c bioconda rnaquast
</code></pre>
</li>
<li>
<p>To verify your installation run</p>
<pre><code> python rnaQUAST.py --test 
</code></pre>
</li>
<li>
<p>To run rnaQUAST on your data use the following command</p>
<pre><code> python rnaQUAST.py \
--transcripts /PATH/TO/transcripts1.fasta /PATH/TO/ANOTHER/transcripts2.fasta /PATH/TO/MULTIPLE/*.fasta [...] \
--reference /PATH/TO/reference_genome.fasta --gtf /PATH/TO/gene_coordinates.gtf
</code></pre>
</li>
</ul>
<p><a name="sec2"></a></p>
<h2>2 Installation &amp; requirements</h2>
<p><a name="sec2.1"></a></p>
<h3>2.1 General requirements</h3>
<p>rnaQUAST can be installed via conda:</p>
<pre><code>    conda install -c bioconda rnaquast
</code></pre>
<p>If you wish to run rnaQUAST from <a href="https://github.com/ablab/rnaquast/releases">the release archive</a> you need:</p>
<ul>
<li>Python3 or Python2 (2.5+)</li>
<li><a href="http://matplotlib.org/" rel="nofollow">matplotlib</a> python package</li>
<li><a href="https://joblib.readthedocs.io/en/latest/" rel="nofollow">joblib</a> python package</li>
<li><a href="https://pythonhosted.org/gffutils/installation.html" rel="nofollow">gffutils</a> python package (needs <a href="http://biopython.org" rel="nofollow">biopython</a>)</li>
<li><a href="http://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/" rel="nofollow">NCBI BLAST+ (blastn)</a></li>
<li><a href="http://research-pub.gene.com/gmap/" rel="nofollow">GMAP</a> (or <a href="http://hgwdev.cse.ucsc.edu/~kent/exe/" rel="nofollow">BLAT</a>) aligner</li>
</ul>
<p>rnaQUAST still works under Python2 (2.5+), but since Python2 is outdated, its support is not maintained since version 2.0.</p>
<p>Note, that due to the limitations of <code>BLAT</code>, in order to work with reference genomes of size more than 4 Gb a <code>pslSort</code> is also required.</p>
<p>Paths to <code>blastn</code> and <code>GMAP</code> (or <code>BLA</code>T) should be added to the <code>$PATH</code> environmental variable. To check that everything is installed correctly we recommend to run:</p>
<pre><code>python rnaQUAST.py --test
</code></pre>
<p>Note that <code>gffutils</code> is used to complete gene coordinates in case of missing transcripts / genes records. For more information, see <a href="#sec3.3">advanced options</a>.<a name="sec2.2"></a></p>
<h3>2.2 Software for <em>de novo</em> quality assessment</h3>
<p>When reference genome and gene database are unavailable, we recommend to run <a href="http://busco.ezlab.org/" rel="nofollow">BUSCO</a> and <a href="http://topaz.gatech.edu/GeneMark/" rel="nofollow">GeneMarkS-T</a> in rnaQUAST pipeline.</p>
<p><strong>BUSCO requirements</strong></p>
<p>BUSCO allows to detect core genes in the assembled transcripts. To use it you should install <a href="http://busco.ezlab.org/" rel="nofollow">BUSCO v4+</a>, <a href="http://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/" rel="nofollow">tblastn</a>, <a href="http://hmmer.janelia.org/" rel="nofollow">HMMER</a> and transeq and add these tools to the <code>$PATH</code> variable.</p>
<p>To run BUSCO provide lineage-specific database name via <code>--busco</code> option. You may also download the appropriate database from <a href="http://busco.ezlab.org" rel="nofollow">http://busco.ezlab.org</a> manually and provide it using the same option (see <a href="#busco">options</a> for details).</p>
<p><strong>GeneMarkS-T requirements</strong></p>
<p><a href="http://topaz.gatech.edu/GeneMark/" rel="nofollow">GeneMarkS-T</a> allows to predict genes in the assembled transcripts without reference genome. If you wish to use it in rnaQUAST pipeline, GeneMarkS-T should be properly installed and added to the <code>$PATH</code> variable.</p>
<p><a name="sec2.3"></a></p>
<h3>2.3 Read alignment software</h3>
<p>rnaQUAST is also capable of calculating various statistics using raw reads (e.g. database coverage by reads). To obtain them you need to install <a href="https://github.com/alexdobin/STAR">STAR</a> aligner and add it to the <code>$PATH</code> variable. To learn more see <a href="#readopts">input options</a>.</p>
<p><a name="sec3"></a></p>
<h2>3 Options</h2>
<p><a name="sec3.1"></a></p>
<h3>3.1 Input data options</h3>
<p>To run rnaQUAST you need to provide either FASTA files with transcripts (recommended), or align transcripts to the reference genome manually and provide the resulting PSL files.</p>
<p><code>-r &lt;REFERENCE&gt;, --reference &lt;REFERENCE&gt;</code><br>
    Single file with reference genome containing all chromosomes/scaffolds in FASTA format (preferably with <code>*.fasta, *.fa, *.fna, *.ffn or *.frn</code> extension) OR<br>
    <strong><code>*.txt</code></strong> file containing the one-per-line list of FASTA files with reference sequences.</p>
<p><code>--gtf &lt;GENE_COORDINATES&gt;</code><br>
    File with gene coordinates in GTF/GFF format (needs information about parent relations). We recommend to use files downloaded from <a href="http://www.gencodegenes.org/" rel="nofollow">GENCODE</a> or Ensembl.</p>
<p><code>--gene_db &lt;GENE_DB&gt;</code><br>
    Path to the gene database generated by gffutils. The database is created during the first run. This option is not compatible with <code>--gtf</code> option. We recommend to use this option once the database is created in order to speed up the run.</p>
<p><code>--gmap_index &lt;INDEX FOLDER&gt;,</code><br>
    Folder containing pre-built GMAP index for the reference genome. Using previously constructed index decreases running time. Note, that you still need to provide the reference genome that was used for index construction when this option is used.</p>
<p><code>-c &lt;TRANSCRIPTS ...&gt;, --transcripts &lt;TRANSCRIPTS, ...&gt;</code><br>
     File(s) with transcripts in FASTA format separated by space. Wildcards can be used, e.g. <code>--transcripts */*.fasta</code>.</p>
<p><code>-psl &lt;TRANSCRIPTS_ALIGNMENT ...&gt;, --alignment &lt;TRANSCRIPTS_ALIGNMENT, ...&gt;</code><br>
     File(s) with transcript alignments to the reference genome in PSL format separated by space.</p>
<p><a name="readopts"></a></p>
<p><code>-sam &lt;READS_ALIGNMENT&gt;, --reads_alignment &lt;READS_ALIGNMENT&gt;</code><br>
     File with read alignments to the reference genome in SAM format.</p>
<p><code>-1 &lt;LEFT_READS&gt;, --left_reads &lt;LEFT_READS&gt;</code><br>
     File with forward paired-end reads in FASTQ or gzip-compressed fastq format.</p>
<p><code>-2 &lt;RIGHT_READS&gt;, --right_reads &lt;RIGHT_READS&gt;</code><br>
     File with reverse paired-end reads in FASTQ or gzip-compressed fastq format.</p>
<p><code>-s &lt;SINGLE_READS&gt;, --single_reads &lt;SINGLE_READS&gt;</code><br>
     File with single reads in FASTQ or gzip-compressed fastq format.</p>
<p><a name="sec3.2"></a></p>
<h3>3.2 Basic options</h3>
<p><code>-o &lt;OUTPUT_DIR&gt;, --output_dir &lt;OUTPUT_DIR&gt;</code><br>
     Directory to store all results. Default is <code>rnaQUAST_results/results_&lt;datetime&gt;</code>.</p>
<p><code>--test</code><br>
     Run rnaQUAST on the test data from the <code>test_data</code> folder, output directory is <code>rnaOUAST_test_output</code>.</p>
<p><code>-d, --debug</code><br>
     Report detailed information, typically used only for detecting problems.</p>
<p><code>-h, --help</code><br>
     Show help message and exit.</p>
<p><a name="sec3.3"></a></p>
<h3>3.3 Advanced options</h3>
<p><code>-t &lt;INT&gt;, --threads &lt;INT&gt;</code><br>
     Maximum number of threads. Default is min(number of CPUs / 2, 16).</p>
<p><code>-l &lt;LABELS ...&gt;, --labels &lt;LABELS ...&gt;</code><br>
     Name(s) of assemblies that will be used in the reports separated by space and given in the same order as files with transcripts / alignments.</p>
<p><code>--prokaryote</code><br>
     Use this option if the genome is prokaryotic.</p>
<p><code>-ss, --strand_specific</code><br>
     Set if transcripts were assembled using strand-specific RNA-Seq data in order to benefit from knowing whether the transcript originated from the + or - strand.</p>
<p><code>--min_alignment &lt;MIN_ALIGNMENT&gt;</code><br>
     Minimal alignment length to be used, default value is 50.</p>
<p><code>--no_plots</code><br>
     Do not draw plots (makes rnaQUAST run a bit faster).</p>
<p><code>--blat</code><br>
     Run with <a href="http://hgwdev.cse.ucsc.edu/~kent/exe/" rel="nofollow">BLAT alignment tool</a> instead of <a href="http://research-pub.gene.com/gmap/" rel="nofollow">GMAP</a>.</p>
<p><a name="busco"></a><br>
<code>--busco</code><br>
     Run <a href="http://busco.ezlab.org/" rel="nofollow">BUSCO tool</a>, which detects core genes in the assembly (see <a href="#sec2">Installation &amp; requirements</a>). Use this option to provide BUSCO database name to use or path to the local database. Also, you can set <code>auto-lineage</code> for automated lineage selection.</p>
<p><code>--gene_mark</code><br>
     Run with <a href="http://topaz.gatech.edu/GeneMark/" rel="nofollow">GeneMarkS-T</a> gene prediction tool. Use <code>--prokaryote</code> option if the genome is prokaryotic.</p>
<p><code>--disable_infer_genes</code><br>
     Use this option if your GTF file already contains genes records, otherwise gffutils will fix it. Note that gffutils may work for quite a long time.</p>
<p><code>--disable_infer_transcripts</code><br>
     Is option if your GTF file already contains transcripts records, otherwise gffutils will fix it. Note that gffutils may work for quite a long time.</p>
<p><code>--lower_threshold</code><br>
     Lower threshold for x-assembled/covered/matched metrics, default: 50%.</p>
<p><code>--upper_threshold</code><br>
     Upper threshold for x-assembled/covered/matched metrics, default: 95%.</p>
<p><a name="sec4"></a></p>
<h2>4 Understanding rnaQUAST output</h2>
<p>In this section we describe metrics, statistics and plots generated by rnaQUAST. Metrics highlighted with <strong><em>bold italic</em></strong> are considered as the most important and are included in the short summary report (<code>short_report.txt</code>).</p>
<p><a name="sec4"></a></p>
<p>For the simplicity of explanation, <em>transcript</em> is further referred to as a sequence generated by the assembler and <em>isoform</em> denotes a sequence from the gene database. Figure below demonstrates how rnaQUAST classifies transcript and isoform sequences using alignment information. <a target="_blank" rel="noopener noreferrer" href="fig1.png"><img src="fig1.png" alt="" style="max-width:100%;"></a></p>
<p><a name="sec4.1"></a></p>
<h3><a name="sec4.1">4.1 Reports</a></h3>
<p>The following text files with reports are contained in <code>comparison_output</code> directory and include results for all input assemblies. In addition, these reports are contained in <code>&lt;assembly_label&gt;_output</code> directories for each assembly separately.</p>
<p><strong><code>database_metrics.txt</code></strong><br>
Gene database metrics.</p>
<ul>
<li><strong><em>Genes</em></strong> / Protein coding genes – number of genes / protein coding genes</li>
<li>Isoforms / Protein coding isoforms – number of isoforms / protein coding isoforms</li>
<li>Exons / Introns – total number of exons / introns</li>
<li>Total / Average length of all isoforms, bp</li>
<li>Average exon length, bp</li>
<li>Average intron length, bp</li>
<li><strong><em>Average</em></strong> / Maximum number of exons per isoform</li>
</ul>

<p><a name="sec4.1"></a><a name="readcov"></a>Coverage by reads. The following metrics are calculated only when <code>--left_reads</code>, <code>--right_reads</code>, <code>--single_reads</code> or <code>--sam</code> options are used (see <a href="#readopts">options</a> for details).</p>
<ul>
<li>Database coverage – the total number of bases covered by reads (in all isoforms) divided by the total length of all isoforms.</li>
<li>x%-covered genes / isoforms / exons – number of genes / isoforms / exons from the database that have at least x% of bases covered by all reads, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default).</li>
</ul>
<p><strong><code>basic_mertics.txt</code></strong><br>
Basic transcripts metrics are calculated without reference genome and gene database.</p>
<ul>
<li><strong><em>Transcripts</em></strong> – total number of assembled transcripts.</li>
<li><strong><em>Transcripts &gt; 500 bp</em></strong></li>
<li>Transcripts &gt; 1000 bp</li>
<li>Average length of assembled transcripts</li>
<li>Longest transcript</li>
<li>Total length</li>
<li>Transcript N50 – a maximal number N, such that the total length of all transcripts longer than N bp is at least 50% of the total length of all transcripts.</li>
</ul>
<p><strong><code>alignment_metrics.txt</code></strong><br>
Alignment metrics are calculated with reference genome but without using gene database. To calculate the following metrics rnaQUAST filters all short partial alignments (see <a href="#sec3.3"><code>--min_alignment</code> option</a>) and attempts to select the best hits for each transcript.</p>
<ul>
<li><strong><em>Transcripts</em></strong> – total number of assembled transcripts.</li>
<li><strong><em>Aligned</em></strong> – the number of transcripts having at least 1 significant alignment.</li>
<li><strong><em>Uniquely aligned</em></strong> – the number of transcripts having a single significant alignment.</li>
<li>Multiply aligned – the number of transcripts having 2 or more significant alignments. Multiply aligned transcripts are stored in <code>&lt;assembly_label&gt;.paralogs.fasta</code> file.</li>
<li>Misassembly candidates reported by GMAP (or BLAT) – transcripts that have discordant best-scored alignment (partial alignments that are either mapped to different strands / different chromosomes / in reverse order / too far away).</li>
<li><strong><em>Unaligned</em></strong> – the number of transcripts without any significant alignments. Unaligned transcripts are stored in <code>&lt;assembly_label&gt;.unaligned.fasta</code> file.</li>
</ul>
<p>Number of assembled transcripts = Unaligned + Aligned = Unaligned + (Uniquely aligned + Multiply aligned + Misassembly candidates reported by GMAP (or BLAT)).</p>
<p>Alignment metrics for non-misassembled transcripts</p>
<ul>
<li><strong><em>Average aligned fraction.</em></strong> Aligned fraction for a single transcript is defined as total number of aligned bases in the transcript divided by the total transcript length.</li>
<li>Average alignment length. Aligned length for a single transcript is defined as total number of aligned bases in the transcript.</li>
<li>Average blocks per alignment. A block is defined as a continuous alignment fragment without indels.</li>
<li>Average block length (see above).</li>
<li><strong><em>Average mismatches per transcript</em></strong> – average number of single nucleotide differences with reference genome per transcript.</li>
<li>NA50 – N50 for alignments.</li>
</ul>
<p><a name="misassemblies"></a><strong><code>misassemblies.txt</code></strong></p>
<ul>
<li><strong><em>Transcripts</em></strong> – total number of assembled transcripts.</li>
<li>Misassembly candidates reported by GMAP (or BLAT) – transcripts that have discordant best-scored alignment (partial alignments that are either mapped to different strands / different chromosomes / in reverse order / too far away).</li>
<li>Misassembly candidates reported by BLASTN – transcripts are aligned to the isoform sequences extracted from the genome using gene database with BLASTN and then transcripts that have partial alignments to multiple isoforms are selected.</li>
<li><strong><em>Misassemblies</em></strong> – misassembly candidates confirmed by both methods described above. Using both methods simultaneously allows to avoid considering misalignments that can be caused, for example, by paralogous genes or genomic repeats. Misassembled transcripts are stored in <code>&lt;assembly_label&gt;.misassembled.fasta</code> file.</li>
</ul>
<p><strong><code>sensitivity.txt</code></strong><br>
Assembly completeness (sensitivity). For the following metrics (calculated with reference genome and gene database) rnaQUAST attempts to select best-matching database isoforms for every transcript. Note that a single transcript can contribute to multiple isoforms in the case of, for example, paralogous genes or genomic repeats. At the same time, an isoform can be covered by multiple transcripts in the case of fragmented assembly or duplicated transcripts in the assembly.</p>
<ul>
<li><strong><em>Database coverage</em></strong> – the total number of bases covered by transcripts (in all isoforms) divided by the total length of all isoforms.</li>
<li>Duplication ratio – total number of aligned bases in assembled transcripts divided by the total number of isoform covered bases. This metric does not count neither paralogous genes nor shared exons, only real overlaps of the assembled sequences that are mapped to the same isoform.</li>
<li>Average number of transcripts mapped to one isoform.</li>
<li><strong><em>x%-assembled genes / isoforms</em></strong>/ exons – number of genes / isoforms / exons from the database that have at least x% captured by a single assembled transcript, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default). 95%-assembled isoforms are stored in <code>&lt;assembly_label&gt;.95%assembled.fasta</code> file.</li>
<li>x%-covered genes / isoforms– number of genes / isoforms from the database that have at least x% of bases covered by all alignments, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default).</li>
<li><strong><em>Mean isoform assembly</em></strong> – assembled fraction of a single isoform is calculated as the largest number of its bases captured by a single assembled transcript divided by its length; average value is computed for isoforms with &gt; 0 bases covered.</li>
<li>Mean isoform coverage – coverage of a single isoform is calculated as the number of its bases covered by all assembled transcripts divided by its length; average value is computed for isoforms with &gt; 0 bases covered.</li>
<li>Mean exon coverage – coverage of a single exon is calculated as the number of its bases covered by all assembled transcripts divided by its length; average value is computed for exons with &gt; 0 bases covered.</li>
<li>Average percentage of isoform x%-covered exons, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default). For each isoform rnaQUAST calculates the number of x%-covered exons divided by the total number of exons. Afterwards it computes average value for all covered isoforms.</li>
</ul>
<p><a href="http://busco.ezlab.org/" rel="nofollow">BUSCO</a> metrics. The following metrics are calculated only when <code>--busco</code> option is used (see <a href="#busco">options</a> for details).</p>
<ul>
<li><strong><em>Complete</em></strong> – percentage of completely recovered genes.</li>
<li><strong><em>Partial</em></strong> – percentage of partially recovered genes.</li>
</ul>
<p><a href="http://topaz.gatech.edu/GeneMark/" rel="nofollow">GeneMarkS-T</a> metrics. The following metrics are calculated when reference and gene database are not provided or <code>--gene_mark</code> option is used (see <a href="#sec3.3">options</a> for details).</p>
<ul>
<li><strong><em>Genes</em></strong> – number of predicted genes in transcripts.</li>
</ul>
<p><strong><code>specificity.txt</code></strong><br>
Assembly specificity. To compute the following metrics we use only transcripts that have at least one significant alignment and are not misassembled.</p>
<ul>
<li><strong><em>Unannotated</em></strong> – total number of transcripts that do not cover any isoform from the database. Unannotated transcripts are stored in <code>&lt;assembly_label&gt;.unannotated.fasta</code> file.</li>
<li><strong><em>x%-matched</em></strong> – total number of transcripts that have at least x% covering an isoform from the database, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default).</li>
<li><strong><em>Mean fraction of transcript matched</em></strong> – matched fraction of a single transcript is calculated as the number of its bases covering an isoform divided by the transcript length; average value is computed for transcripts with &gt; 0 bases matched.</li>
<li>Mean fraction of block matched – matched fraction of a single block is calculated as the number of its bases covering an isoform divided by the block length; average value is computed for blocks with &gt; 0 bases matched.</li>
<li>x%-matched blocks – percentage of blocks that have at least x% covering an isoform from the database, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default).</li>
<li>Matched length – total number of transcript bases covering isoforms from the database.</li>
<li>Unmatched length – total alignment length - Matched length.</li>
</ul>
<p><strong><code>relative_database_coverage.txt</code></strong><br>
Relative database coverage metrics are calculated only when raw reads (or read alignments) are provided. rnaQUAST uses read alignments to estimate the upper bound of the database coverage and the number of x-covered genes / isoforms / exons (see <a href="#readcov">read coverage</a>) and computes the following metrics:</p>
<ul>
<li><strong><em>Relative database coverage</em></strong> – ratio between transcripts database coverage and reads database coverage.</li>
<li>Relative x%-assembled genes / isoforms / exons – ratio between transcripts x%-assembled and reads x%-covered genes / isoforms / exons.</li>
<li>Relative x%-covered genes / isoforms / exons – ratio between transcripts x%-covered and reads x%-covered genes / isoforms / exons.</li>
</ul>
<p><a name="sec4.2"></a></p>
<h3>4.2 Detailed output</h3>
<p>These files are contained in <code>&lt;assembly_label&gt;_output</code> directories for each assembly separately.</p>
<ul>
<li><code>&lt;assembly_label&gt;.unaligned.fasta</code> – transcripts without any significant alignments.</li>
<li><code>&lt;assembly_label&gt;.paralogs.fasta</code> – transcripts having 2 or more significant alignments.</li>
<li><code>&lt;assembly_label&gt;.misassembled.fasta</code> – misassembly candidates detected by methods described above. See <a href="#misassemblies"><code>misassemblies.txt</code></a> description for details.</li>
<li><code>&lt;assembly_label&gt;.correct.fasta</code> – transcripts with exactly 1 significant alignment that do not contain misassemblies.</li>
<li><code>&lt;assembly_label&gt;.x%-assembled.list</code> – IDs of the isoforms from the database that have at least x% captured by a single assembled transcript, where x is specified by the user with an option <code>--upper_threshold</code> (95% by default).</li>
<li><code>&lt;assembly_label&gt;.unannotated.fasta</code> – transcripts that do not cover any isoform from the database.</li>
</ul>
<p>The following text file is contained in <code>comparison_output</code> directory and <code>&lt;assembly_label&gt;_output</code> directories for each assembly separately.</p>
<ul>
<li><code>reads.x%-covered.list</code> – IDs of the isoforms from the database that have at least x% bases covered by all reads, where x is specified with <code>--lower_threshold / --upper_threshold</code> options (50% / 95% by default).</li>
</ul>
<p><a name="sec4.3"></a></p>
<h3>4.3 Plots</h3>
<p>The following plots are similarly contained in both <code>comparison_output</code> directory and <code>&lt;assembly_label&gt;_output</code> directories. Please note, that most of the plots represent cumulative distributions and some plots are given in logarithmic scale.</p>
<p><strong>Basic</strong></p>
<ul>
<li><strong><em><code>transcript_length.png</code></em></strong> – assembled transcripts length distribution (+ database isoforms length distribution).</li>
<li><code>block_length.png</code> – alignment blocks length distribution (+ database exons length distribution).</li>
<li><code>x-aligned.png</code> – transcript aligned fraction distribution.</li>
<li><code>blocks_per_alignment.png</code> – distribution of number of blocks per alignment (+ distribution of number of database exons per isoform).</li>
<li><code>alignment_multiplicity.png</code> – distribution for the number of significant alignment for each multiply-aligned transcript.</li>
<li><strong><em><code>mismatch_rate.png</code></em></strong> – substitution errors per alignment distribution.</li>
<li><code>Nx.png</code> – Nx plot for transcripts. Nx is a maximal number N, such that the total length of all transcripts longer than N bp is at least x% of the total length of all transcripts.</li>
<li><code>NAx.png</code> – Nx plot for alignments.</li>
</ul>
<p><strong>Sensitivity</strong></p>
<ul>
<li><strong><em><code>x-assembled.png</code></em></strong> – a histogram in which each bar represents the number of isoforms from the database that have at least x% captured by a single assembled transcript.</li>
<li><code>x-covered.png</code> – a histogram in which each bar represents the number of isoforms from the database that have at least x% of bases covered by all alignments.</li>
<li><code>x-assembled_exons.png</code> – a histogram in which each bar represents the number of exons from the database that have at least x% captured by a single assembled transcript.</li>
<li><code>x-covered_exons.png</code> – a histogram in which each bar represents the number of exons from the database that have at least x% of bases covered by all alignments.</li>
<li><code>alignments_per_isoform.png</code> – plot showing number of transcript alignments per isoform</li>
</ul>
<p><strong>Specificity</strong></p>
<ul>
<li><code>x-matched.png</code> – a histogram in which each bar represents the number of transcripts that have at least x% matched to an isoform from the database.</li>
<li><code>x-matched_blocks.png</code> – a histogram in which each bar represents the number of all blocks from all transcript alignments that have at least x% matched to an isoform from the database.</li>
</ul>
<p><a name="sec5"></a></p>
<h2>5 Citation</h2>
<p><a href="https://academic.oup.com/bioinformatics/article/32/14/2210/1743439" rel="nofollow">Bushmanova, E., Antipov, D., Lapidus, A., Suvorov, V. and Prjibelski, A.D., 2016. rnaQUAST: a quality assessment tool for de novo transcriptome assemblies. Bioinformatics, 32(14), pp.2210-2212.</a></p>
<p><a name="sec6"></a></p>
<h2>6 Feedback and bug reports</h2>
<p>Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve rnaQUAST. If you have any troubles running rnaQUAST, please send us <code>logs/rnaQUAST.log</code> from the output directory. Address for communications: <a href="mailto:rnaquast_support@ablab.spbau.ru">rnaquast_support@ablab.spbau.ru</a>.<br>
You may also submit your issue to our <a href="https://github.com/ablab/rnaquast/issues">GitHub repository</a>.</p><br/><br/><br/><br/><br/> </body> </html>
